doc/api/esm.md

```diff
@@ -313,50 +275,349 @@ If a package has no exports, setting `"exports": false` can be used instead of
+certain conditions. They are supported for both CommonJS and ES module imports.
+
+For example, a package that wants to provide different ES module exports for
+Node.js and the browser can be written:
 
 <!-- eslint-skip -->
 ```js
+// ./node_modules/pkg/package.json
 {
+  "type": "module",
+  "main": "./index.js",
   "exports": {
-    "./submodule": ["not:valid", "./submodule.js"]
+    "./feature": {
+      "browser": "./feature-browser.js",
+      "node": "./feature-server.js"
+    }
   }
 }
 ```
 
-Since `"not:valid"` is not a supported target, `"./submodule.js"` is used
-instead as the fallback, as if it were the only target.
+When an exports target condition object is provided like the above, the first
+matching key in the object that is supported by the current environment will
+be resolved.
+
+Node.js will match the `"node"` condition, resulting in `pkg/feature-server.js`
+being used. Browser tooling with support for exports can then resolve the
+`"browser"` condition.
+
+The conditions supported in Node.js are matched in the following order:
+
+1. `"module"` - matched when the package is imported as an ES module.
+2. `"node"` - matched for any Node.js environment. Can be a CommonJS or ES
+   module file.
+3. `"default"` - the generic fallback that will always match if no other
+   more specific condition is matched first. Can be a CommonJS or ES module
+   file.
+
+Using the `"module"` condition it is possible to define a package that will
+have a different exported value for CommonJS and ES modules, which can be a
+hazard in that it can result in having two separate instances of the same
+package in use in an application, which can cause a number of bugs.
+
+Other conditions such as `"browser"`, `"electron"`, `"deno"`, `"react-native"`
+etc. could be defined in other runtimes or tools.
+
+### Dual CommonJS/ES Module Packages
+
+Prior to the introduction of support for ES modules in Node.js, it was a common
+pattern for package authors to include both CommonJS and ES module JavaScript
+sources in their package, with `package.json` `"main"` specifying the CommonJS
+entry point and `package.json` `"module"` specifying the ES module entry point.
+This enabled Node.js to run the CommonJS entry point while build tools such as
+bundlers used the ES module entry point, since Node.js ignored (and still
+ignores) the top-level `"module"` field.
+
+Node.js can now run ES module entry points, and using [conditional exports][] it
+is possible to define separate package entry points for CommonJS and ES module
+consumers. Unlike in the scenario where `"module"` is only used by bundlers, or
+ES module files are transpiled into CommonJS on the fly before evaluation by
+Node.js, the files referenced by the ES module entry point are evaluated as ES
+modules.
+
+#### Divergent Specifier Hazard
+
+When one application is using both CommonJS and ES module code provided by one
+package, there is a risk of certain bugs that only occur under these conditions.
```

Comment:
I will think about this and try to give a proper response soon.

I also welcome input from others, wording is not exactly my strength.

Reply:
> When one application is using both CommonJS and ES module code provided by one package, there is an added risk of bugs caused by loading multiple instances of singletons.

This is the best I can come up with right now.  Maybe someone can think of coherent way for this place to also directly mention duplicated installation due to conflicting versions but just calling it an "added risk" rather than the sole cause will make me happy enough.